java/WordCountOptimized.java

```diff
@@ -217,33 +207,66 @@ public static int tokenizeAndSubmitBlock(TObjectIntCustomHashMap<byte[]> counts,
         }
         return 0;
     }
-    
-    private static void submitWord(TObjectIntCustomHashMap<byte[]> m, byte[] word) {
-        m.adjustOrPutValue(word, 1, 1);        
+
+    private static final int TABLE_SIZE = 1024 * 1024;
+    private static final int TABLE_MASK = TABLE_SIZE - 1;
+
+    static class WordMap {
+        final byte[][] keys = new byte[TABLE_SIZE][];
+        final int[] values = new int[TABLE_SIZE];
+        int size = 0;
     }
-    
+
+    private static int submitWord(byte[][] keys, int[] values, int size,
+                                   byte[] word, int off, int len) {
+        int hashCode = (int) XX_HASH.hashBytes(word, off, len);
+        int index = hashCode & TABLE_MASK;
```

Comment:
This means this is not word count, because there shouldn't be more than 500k words in any language. (even with forms, not 25 mln).

I agree on a small input pre-alloc of multi-million table will be slower. But for optimization of a particular use case it's better to pre-alloc it so there are no rehashes.

You shouldn't explain to me how write hash tables :) Do you realize that I'm the author of koloboke?